state-chain/custom-rpc/src/broker.rs

```diff
@@ -191,35 +228,49 @@ where
 						dca_parameters,
 					},
 				),
-				wait_for_finality.unwrap_or_default(),
 				true,
 			)
 			.await
 			.map_err(CfApiError::from)?;
 
-		Ok(extract_from_first_matching_event!(
-			events,
-			cf_static_runtime::swapping::events::SwapDepositAddressReady,
-			{
-				deposit_address,
-				channel_id,
-				source_chain_expiry_block,
-				channel_opening_fee,
-				refund_parameters
+		// Get the pre-allocated channels from the previous finalized block
+		let source_chain: ForeignChain = source_asset.into();
+		let pre_allocated_channels = self.rpc_backend.with_runtime_api(
+			Some(self.rpc_backend.client.info().finalized_hash),
+			|api, hash| {
+				api.cf_get_preallocated_deposit_channels(
+					hash,
+					self.signed_pool_client.account_id(),
+					source_chain,
+				)
 			},
-			SwapDepositAddress {
-				address: AddressString::from_encoded_address(deposit_address.0),
-				issued_block: header.number,
-				channel_id,
-				source_chain_expiry_block: source_chain_expiry_block.into(),
-				channel_opening_fee: channel_opening_fee.into(),
-				refund_parameters: ChannelRefundParameters::from(refund_parameters)
-				.map_address(|refund_address| {
-					AddressString::from_encoded_address(&refund_address.0)
-				}),
+		)?;
+
+		while let Some(status) = status_stream.next().await {
+			match status {
+				TransactionStatus::InBlock((block_hash, tx_index)) => {
+					let swap_deposit_address =
+						self.extract_swap_deposit_address(block_hash, tx_index).await?;
+
+					// If the extracted deposit channel was pre-allocated to this broker
+					// in the previous finalized block, we can return it immediately.
+					// Otherwise, we need to wait for the transaction to be finalized.
+					if pre_allocated_channels
+						.iter()
+						.any(|channel_id| channel_id == &swap_deposit_address.channel_id)
+					{
+						return Ok(swap_deposit_address);
+					}
+				},
+				TransactionStatus::Finalized((block_hash, tx_index)) =>
+					return self.extract_swap_deposit_address(block_hash, tx_index).await,
+				_ =>
+					if let Some(e) = is_transaction_status_error(&status) {
```

Comment:
This is a bit of a strange construction to me. We return an Error inside a Some, meaning we have to unpack the option to throw the Error. 
If `is_transaction_status_error` returns a `Result<(),PoolClientError>` instead, it simplifies a lot of this. Here we can do `.map_err()` and in the other places we can simply use `?` directly.